README.md
# Nemini 
### '/nɛmɪnaɪ/' _Nem-in-eye_  
  
## The primary git repo for Nemini is on Codeberg - please use that repo - https://codeberg.org/pswilde/nemini  

[![Build Status](https://nimble.directory/ci/badges/nemini/nimdevel/status.svg)](https://nimble.directory/ci/badges/nemini/nimdevel/output.html)
[![Build Status](https://nimble.directory/ci/badges/nemini/nimdevel/docstatus.svg)](https://nimble.directory/ci/badges/nemini/nimdevel/doc_build_output.html)

Web : https://paulwilde.uk/dev/nemini  
Gemini : gemini://paulwilde.uk/dev/nemini  
A simple (to configure) Gemini server capable of serving static files with virtual host support.  
Nemini also has a basic header/footer implementation - you can create a header and footer .gemini file in your root which will be applied to each page!  
Leveraging the good work done by @benob for the [Nim Gemini Library](https://github.com/benob/gemini)

## Features
* Serves static files
* Has virtual host capability with aliases
* Header and Footer can be applied to each page

## Installation
Nemini can be installed in a variety of ways. The processes below set Nemini up in such a way that it should almost 'just work'™ (i.e. copy folders system-wide, create nemini user, create systemd service, etc.)

### Arch Linux
Nemini is available in the AUR and can be installed via yay or your desired AUR package manager  
```sh
yay -S nemini-git
```
or  
```sh
git clone https://aur.archlinux.org/nemini-git.git
cd nemini-git
makepkg -si
```

### From Source
Nemini can be easily compiled from source within a few minutes.  
```sh
git clone https://codeberg.org/pswilde/nemini.git
cd nemini
make              # Will build the binary, effectively just runs `nimble build`
sudo make install # This will copy files and folders to system-wide locations
```

## How to run
* Edit the `/etc/nemini/nemini.toml` file as you need.
* Certificates are **REQUIRED**, so make sure you have those set in your config. (Automatically created if they don't exist)
* Host your gemini files in the desired `root_dir` directory
  
If Nemini has been installed via the above Installation methods, then `systemctl enable --now nemini` should start the service and get your server running. Otherwise you can run the binary directly if you have permission on the folders in your config file.

### Parameters
Passing `--config` parameter with a config file location will allow the use of a custom config file saved wherever you wish.

## Certificates
With Gemini, Certificates are mandatory. 
If certificates don't exist they will be created via openssl when first run.

## Static files
Nemini is able to serve static gemtext files from your root directory or subdirectories.  
Your static files must use the extension `.gemini`, `.gmi` or `.gmni`. Currently, other file extensions are not supported, but easily fixed if there's need for it. 

## Headers and Footers
In your sites `root_dir`, add the files `header.gemini` and `footer.gemini` and they will be prepended and appended to your site content

## Contact me
Say "Hi"!  
I'm on the Fediverse at [@paul@notnull.click](https://notnull.click/paul)  
Or raise an issue here (Codeberg) for anything related to issues with Nemini


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
## [Unrelease

### Added
### Fixed

## [0.2.6]

### Added
- added extra line if a footer exists to be placed before the footer (so content doesn't mush up)
- added redirect functionality - make the first page of the content 30 or 31 followed by the url to redirect to
- some more descriptive startup messages
### Fixed

## [0.2.4]

### Added
- Created better defaults for certificate. Also can be changed in config
- Can use a url with a directory name which will return an \_index.gemini/gmi/gmni file if in place

### Fixed

## [0.2.3]

### Added
- headers and footers!

### Fixed
- will now generate certs with alt names
- Error messages a little better
- Issue where pages will always show even if they don't exist due to the header and footer
### Removed

## [0.2.2]

### Added
- -v or --version flags to show current version
### Fixed
### Removed

## [0.2.0]

### Added 
- changed config file format - can now have multiple vhosts on one port
- added --config parameter to load custom config at start. If not set, nemini will check in a few places for one and if still not found will use the sample config

### Fixed
- Lowered required Nim version to 1.6.0
- Removed ability to open any file extension if given in the url (concerns about security). Can now only serve .gemini, .gmi or .gmni files  

## [0.1.0]

### Added
- First release
- Will serve static files only
- Certificates generated on start up unless they already exist


src/config.nim
import os,strutils
import parsetoml

type
  Nemini* = object
    version*: string
    listeners*: seq[Listener]
  Listener* = object
    port*: int
    sites*: seq[Site]
    cert*: Cert
  Cert* = object
    fullchain*: string
    private_key*: string
    days*: int
    country*: string
    state*: string
    locality*: string
    organization*: string
  Site* = object
    name*: string
    base_url*: string
    aliases*: seq[string]
    root_dir*: string
    index*: string

proc getVersion(): string =
  # Get the version number from nimble file
  # TODO probably a better way to do this but it works
  let c = staticRead("../nemini.nimble")
  let l = c.find("version")
  let e = c.find("=",l)
  let cr = c.find("\n",e)
  let v = c[e+1..cr-1]
  let version = v.strip(chars = {' ','"'})
  return version

const VERSION = getVersion()

proc newNemini*(): Nemini =
  return Nemini(version: VERSION)

proc newSite(): Site =
  return Site(index: "index.gemini")

proc newCert(): Cert =
  var c = Cert()
  c.days = 365
  c.country = "GB"
  c.state = "myState"
  c.locality = "myCity"
  c.organization = "myOrganizationName"
  return c

proc newListener(): Listener =
  return Listener(port: 1965, cert: newCert())

proc getSiteConfig(toml: TomlValueRef): Site =
  var site = newSite()
  site.name = toml.getOrDefault("name").getStr
  site.base_url = toml.getOrDefault("base_url").getStr
  if toml.hasKey("aliases"):
    for x in toml["aliases"].getElems:
      site.aliases.add(x.getStr)
  site.root_dir = toml.getOrDefault("root_dir").getStr
  if site.root_dir.contains("{pwd}"):
    site.root_dir = site.root_dir.replace("{pwd}",getCurrentDir())
  return site

proc getNeminiConfig*(file: string): Nemini =
  var nemini = newNemini()
  var cfg = "./config/nemini.sample.toml"
  if file != "":
    cfg = file
  else:
    if fileExists("/etc/nemini/nemini.toml"):
      cfg = "/etc/nemini/nemini.toml"
    elif fileExists("./config/nemini.toml"):
      cfg = "./config/nemini.toml"
  echo "Using config from : ", cfg
  if not fileExists(cfg):
    echo "Config file does not exist"
    quit(1)
  let toml = parsetoml.parseFile(cfg)
  for l in toml["listeners"].getElems:
    var listener = newListener()
    if l.hasKey("cert"):
      let cert = l["cert"].getTable
      listener.cert.fullchain = cert.getOrDefault("fullchain").getStr
      listener.cert.private_key = cert.getOrDefault("private_key").getStr
      if cert.hasKey("days"):
        listener.cert.days = cert["days"].getInt
      if cert.hasKey("country"):
        listener.cert.country = cert["country"].getStr
      if cert.hasKey("state"):
        listener.cert.state = cert["state"].getStr
      if cert.hasKey("locality"):
        listener.cert.locality = cert["locality"].getStr
      if cert.hasKey("organization"):
        listener.cert.organization = cert["organization"].getStr
    if l.hasKey("port"):
      listener.port = l.getOrDefault("port").getInt
    for s in l["sites"].getElems:
      listener.sites.add(getSiteConfig(s))
    if listener.cert.fullchain == "":
      listener.cert.fullchain = "certs/" & listener.sites[0].base_url & ".cert"
    if listener.cert.private_key == "":
      listener.cert.private_key = "certs/" & listener.sites[0].base_url & ".key"
    nemini.listeners.add(listener)
  return nemini



src/certificates.nim
import os,sugar,osproc,sequtils
import config

proc createCerts*(l: Listener): bool =
  createDir("certs")
  let base_site = l.sites[0]
  echo "Creating certificates for ", base_site.name
  let cn = base_site.base_url
  var alt_names = collect:
    for an in base_site.aliases:
      an
  let alt_sites_cn = collect:
    for n in l.sites:
      n.base_url
  alt_names = alt_names.concat(alt_sites_cn)
  let alt_sites_alt_names = collect:
    for n in l.sites:
      for an in n.aliases:
        an
  alt_names = alt_names.concat(alt_sites_alt_names)
  echo alt_names
  var cmd = "openssl req -new -newkey rsa:4096 -days " & $l.cert.days & " "
  cmd    &= "-subj \"/C=" & l.cert.country & "/ST=" & l.cert.state & "/L=" & l.cert.locality & "/O=" & l.cert.organization & "/CN=" & cn & "\" "
  if len(alt_names) > 0:
    var san = "-addext \"subjectAltName="
    for n in alt_names:
      san  &= "DNS:" & n & ", "
    san = san[0 .. ^3]
    san &= "\" "
    cmd &= san
  cmd    &= "-nodes -x509 -keyout " & l.cert.private_key & " -out " & l.cert.fullchain
  let output = execCmd(cmd)
  return output == 0

proc hasCerts*(l: Listener): bool =
  if fileExists(l.cert.fullchain) and fileExists(l.cert.private_key):
    echo "Listener has cert : ", l.port
    return true
  else:
    return createCerts(l)
  return false


src/nemini.nim
import asyncdispatch,os,sugar,strutils
import config
import certificates
import gemini
import simpleargs

proc parseArgs(): ParsedValues =
  var p: ParseSchema
  p.initParser("Nemini - A simple Gemini server"):
    p.addOption(long="--config", help="path to config file")#, default="/etc/nemini/nemini.toml")
    p.addFlag("-v", "--version", help="shows the current nemini version")#, default="/etc/nemini/nemini.toml")
    let opts = p.parseOptions()
    if opts.isSet("v") or opts.isSet("version"):
      echo "Nemini Version : ", newNemini().version
      quit(0)
    return opts

let opts = parseArgs()
let nemini = getNeminiConfig(opts["config"])

proc checkStatusCodes(content: string): (int, string) =
  if content == "":
    # Temporary failure
    return (51,content)
  case content[0]:
    of '3':
      let code = content[0..1].parseInt
      # Redirect - 30 == temporary, 31 == permanent
      return (code,content[3..^2])
    else:
      # There's content, return it
      return (20,content)

proc addHeaderAndFooter(content: var string, site: Site): string =
  let header = site.root_dir / "header.gemini"
  let footer = site.root_dir / "footer.gemini"
  if fileExists(header):
    content = readFile(header) & content
  if fileExists(footer):
    content &= "\r\n"
    content = content & readFile(footer)
  return content

proc getContent(s: Site, path: string): string =
  const extensions = [".gemini", ".gmi", ".gmni"]
  var p = path
  if path == "" or path == "/":
    p = s.index
  let page = s.root_dir / p
  var content = ""
  for ext in extensions:
    # If the given url already has an extension
    if ext in page:
      content &= readFile(page)
    # otherwise check it's a valid one
    let pe = page & ext
    if fileExists(pe):
      content &= readFile(pe)
    # if still nothing, check to see if there's an 
    # _index/.ge?m(in)?i/ file in the page named directory
    let pdi = page / "_index" & ext
    if fileExists(pdi):
      content &= readFile(pdi)
  if content == "":
    return content

  # Removed for security - could potentially open any file on file system
  #if fileExists(page):
  #  return readFile(page)
  
  return content

proc findSite(url: string): Site =
  # TODO probably could do this better, ideally passing the listener into the handle procedure may make this a little nicer
  let sites = collect:
    for l in nemini.listeners:
      for s in l.sites:
        if s.base_url == url or s.aliases.contains(url):
          s
  return sites[0]

proc handle(req: AsyncRequest) {.async.} =
  try:
    # TODO go and find the site - not ideal, need to work out how to send the site the callback
    let site = findSite(req.url.hostname)
    # TODO if a file exists in the root directory use it, if not show an error
    var content = site.getContent(req.url.path)
    let (status,meta) = content.checkStatusCodes()
    content = content.addHeaderAndFooter(site)
    let furl = req.url.hostname / req.url.path
    case status:
      of 20:
        echo "Found : ", req.url.hostname / req.url.path
        await req.respond(Success, "text/gemini", content)
      of 30:
        echo "TempRedirect : ", furl , " to ", meta
        await req.respond(Redirect, meta)
      of 31:
        echo "TempRedirect : ", furl , " to ", meta
        await req.respond(Redirect, meta)
      of 51:
        echo "NotFound : ", furl
        await req.respond(NotFound, "Page Not Found.")
      else:
        echo "Server Unavailable : ", furl
        await req.respond(ServerUnavailable, "Unknown Error")
  except:
    echo "ERROR: " & getCurrentExceptionMsg()
    await req.respond(ERROR, "Server Error")


when isMainModule:
  if len(nemini.listeners) > 0:
    echo "Starting Nemini version : " & nemini.version
    for l in nemini.listeners:
      echo "Starting listener on port : ", l.port
      if l.hasCerts() and len(l.sites) > 0:
        try:
          var server = newAsyncGeminiServer(certFile = l.cert.fullchain, keyFile = l.cert.private_key)
          # TODO send the listener into the handle callback for less work finding the site later
          asyncCheck server.serve(Port(l.port), handle)
          echo "Started on port : ", l.port
        except:
          echo "Failed to open certificate - do you have permission?"
      else:
        echo "No Certificate found for Listener : ", l.port, ". Not Starting."
    runForever()
